h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2200,6 +2339,238 @@ private void fitIRLSM(Solver s) {
         Log.warn(LogMsg("Got Non SPD matrix, stopped."));
       }
     }
+    
+    /***
+     * This method fits the constraint GLM for IRLSM.  Please refer to doc: 
+     */
+    private void fitIRLSMCS(Solver s) {
+      double[] betaCnd = _checkPointFirstIter ? _model._betaCndCheckpoint : _state.beta();
+      double[] tempBeta = _parms._separate_linear_beta ? new double[betaCnd.length] : null;
+      List<String> coeffNames = Arrays.stream(_state.activeData()._coefNames).collect(Collectors.toList());
+
+      final BetaConstraint bc = _parms._separate_linear_beta ? _state.activeBC() : null;
+      ExactLineSearch ls = null;
+      int iterCnt = (_checkPointFirstIter ? _state._iter : 0)+_initIter;
+      _state.setConstraintInfo(betaCnd, coeffNames);  // setup the constraints
+      _state.initConstraintInfo(coeffNames);  // setup constraints derivatives and double derivatives
+      GLMGradientSolver ginfo = gam.equals(_parms._glmType) ? new GLMGradientSolver(_job, _parms, _dinfo, 0,
+              _state.activeBC(), _betaInfo, _penaltyMatrix, _gamColIndices) : new GLMGradientSolver(_job, _parms, 
+              _dinfo, 0, _state.activeBC(), _betaInfo);
+      GLMGradientInfo gradientInfo = calGradient(betaCnd, _state, ginfo);
+      _state.setGradientInfo(gradientInfo);  // update state ginfo with contributions from GLMGradientInfo
+      boolean predictorSizeChange;
+      boolean applyBetaConstraints = _parms._separate_linear_beta && _betaConstraintsOn;
+      // short circuit check here: if gradient magnitude is small and all constraints are satisfied, quit right away
+      if (constraintsStop(gradientInfo, _state)) {
+        Log.info(LogMsg("GLM with constraints model building completed successfully!!"));
+        return;
+      }
+      double gradMagSquare;
+      int origIter = iterCnt+1;
+      try {
+        while (true) {
+          do {
+            iterCnt++;
+            long t1 = System.currentTimeMillis();
+            ComputationState.GramGrad gram = _state.computeGram(betaCnd, gradientInfo);  // calculate gram (hessian), xy, objective values
+            if (iterCnt == origIter) {
+              Matrix gramMatrix = new Matrix(gram._gram);
+              if (gramMatrix.cond() >= BAD_CONDITION_NUMBER)
+                if (_parms._init_optimal_glm) {
+                  warn("init_optimal_glm", " should be disabled.  This lead to gram matrix being close to" +
+                          " singular.  Please re-run with init_optimal_glm set to false.");
+                  Log.warn("init_optimal_glm", " should be disabled.  This lead to gram matrix being close to" +
+                          " singular.  Please re-run with init_optimal_glm set to false.");
```

Comment:
This seem unnecessary. `ModelBuilder.warn` should log as well (in ValidationMessage constructor):

https://github.com/h2oai/h2o-3/blob/5927bb1ec7ca75037c350cb9058611c8cfcfa3a8/h2o-core/src/main/java/hex/ModelBuilder.java#L1332-L1388

Reply:
Removed.